packages/synapse-react-client/src/components/DataGrid/SynapseGrid.tsx

```diff
@@ -232,17 +172,55 @@ const SynapseGrid = forwardRef<
     operations = removeNoOpOperations(newValue, rowValues, operations)
 
     if (operations.length > 0) {
-      // Apply the changes to the model
-      applyOperationsToModel(operations, newValue, rowValues, model)
+      // Track row creation, updates, and deletions to keep UI state and undo history in sync
+
+      // Add all operations to the undo stack
+      addOperationsToUndoStack(operations, rowValues, newValue)
+
+      // Transform operations to model changes
+      const modelChanges = mapOperationsToModelChanges(
+        operations,
+        newValue,
+        rowValues,
+      )
+
+      // Apply each change to the model
+      modelChanges.forEach(change => {
+        applyModelChange(model, change)
+      })
 
       // Push the changes to the server (throttled)
       commit()
     }
   }
 
+  const applyModelChangeFromUndo = useCallback(
+    (change: ModelChange) => {
+      if (!model) {
+        console.error('Model is not initialized')
+        return
+      }
+
+      if (change.type === 'DELETE' && gridRef.current) {
+        // The user may have set a cell as active that we are removing with an 'undo'. In that case, clear the active state
+        gridRef.current.setActiveCell(null)
+      }
+
+      applyModelChange(model, change)
+      commit()
```

Comment:
Very minor improvement - you could refactor out a function
```ts
function applyAndCommitChanges(model: Model, changes: ModelChange[]) {
  modelChanges.forEach(change => {
    applyModelChange(model, change)
  })
  commit()
}
```` 

that is called by both `applyModelChangeFromUndo` and `handleChange`